File: src/main/java/pl/ogiba/dwarf/utils/JsonConverter.java
package pl.ogiba.dwarf.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javafx.util.StringConverter;

/**
 *
 * @author robertogiba
 */
public class JsonConverter extends StringConverter<String> {

    private static final int DEFUALT_INDENT = 4;

    @Override
    public String toString(String object) {
        if (object == null) {
            return "";
        }

        return object;
    }

    @Override
    public String fromString(String string) {
        String formattedValue = formatJson(string, DEFUALT_INDENT);

        return formattedValue;
    }

    public String toPrettyFormat(String jsonString) {
        JsonParser parser = new JsonParser();
        JsonElement jsonElement = parser.parse(jsonString);

        Gson gson = new GsonBuilder().setPrettyPrinting().create();

        String prettyJson;
        if (jsonElement.isJsonArray()) {
            JsonArray json = jsonElement.getAsJsonArray();

            prettyJson = gson.toJson(json);
        } else {
            JsonObject json = jsonElement.getAsJsonObject();

            prettyJson = gson.toJson(json);
        }

        return prettyJson;
    }

    public String formatJson(String jsonString, int indent) {
        Pattern regex = Pattern.compile("[\\[\\{\"*\",\\}\\]]");

        int nodeDepth = 0;
        boolean openQuotation = true;

        StringBuilder builder = new StringBuilder();
        for (char singleChar : jsonString.toCharArray()) {
            String currentChar = String.valueOf(singleChar);

            final Matcher matcher = regex.matcher(currentChar);

            if (matcher.find()) {
                switch (currentChar) {
                    case "]":
                        builder.append("\n").append(currentChar);
                        break;
                    case "{":
                    case "[":
                        final boolean isColon = findColon(builder.toString());

                        if (!isColon) {
                            builder.append(indentValue(indent * nodeDepth));
                        }

                        nodeDepth++;
                        builder.append(currentChar).append("\n");
                        openQuotation = true;
                        break;
                    case "\"":
                        if (openQuotation) {
                            openQuotation = false;
                            builder.append(indentValue(indent * nodeDepth));
                        }

                        builder.append(currentChar);
                        break;
                    case "}":
                        nodeDepth--;
                        builder.append("\n")
                                .append(indentValue(indent * nodeDepth))
                                .append(currentChar);
                        break;
                    case ",":
                        builder.append(currentChar)
                                .append("\n");
                        openQuotation = true;
                        break;
                    default:
                        builder.append(indentValue(indent * nodeDepth))
                                .append(currentChar)
                                .append("\n");
                        break;
                }
            } else {
                builder.append(currentChar);
            }
        }
        return builder.toString();
    }

    private boolean findColon(String valueToCheck) {
        if (valueToCheck.length() < 1) {
            return false;
        }
        char[] currentFormattedChars = valueToCheck.trim()
                .toCharArray();
        final int lastIndex = currentFormattedChars.length - 1;

        return currentFormattedChars[lastIndex] == ':';
    }

    private String indentValue(int depth) {
        String indentedValue = "";
        for (int j = 0; j < depth; j++) {
            indentedValue += " ";
        }
        return indentedValue;
    }
}

File: src/main/java/pl/ogiba/dwarf/utils/JsonConverter.kt
package pl.ogiba.dwarf.utils

import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import javafx.util.StringConverter
import java.util.regex.Pattern

/**
 *
 * @author robertogiba
 */
class JsonConverter : StringConverter<String>() {

    override fun toString(`object`: String?) = `object` ?: ""

    override fun fromString(string: String) = formatJson(string, DEFAULT_INDENT)

    fun toPrettyFormat(jsonString: String): String {
        val parser = JsonParser()
        val jsonElement = parser.parse(jsonString)
        val gson = GsonBuilder().setPrettyPrinting().create()

        return if (jsonElement.isJsonArray) {
            val json = jsonElement.asJsonArray
            gson.toJson(json)
        } else {
            val json = jsonElement.asJsonObject
            gson.toJson(json)
        }
    }

    fun formatJson(jsonString: String, indent: Int): String {
        val regex = Pattern.compile("[\\[{\"*\",}\\]]")

        var nodeDepth = 0
        var openQuotation = true

        return StringBuilder().let { builder ->
            for (singleChar in jsonString.toCharArray()) {
                val currentChar = singleChar.toString()

                val matcher = regex.matcher(currentChar)

                if (matcher.find()) {
                    when (currentChar) {
                        "]"      -> builder.append("\n").append(currentChar)
                        "{", "[" -> {
                            val isColon = findColon(builder.toString())

                            if (!isColon) {
                                builder.append(indentValue(indent * nodeDepth))
                            }

                            nodeDepth++
                            builder.append(currentChar).append("\n")
                            openQuotation = true
                        }
                        "\""     -> {
                            if (openQuotation) {
                                openQuotation = false
                                builder.append(indentValue(indent * nodeDepth))
                            }

                            builder.append(currentChar)
                        }
                        "}"      -> {
                            nodeDepth--
                            builder.append("\n").append(indentValue(indent * nodeDepth)).append(currentChar)
                        }
                        ","      -> {
                            builder.append(currentChar).append("\n")
                            openQuotation = true
                        }
                        else     -> builder.append(indentValue(indent * nodeDepth)).append(currentChar).append("\n")
                    }
                } else {
                    builder.append(currentChar)
                }
            }
            builder
        }.toString()
    }

    private fun findColon(valueToCheck: String): Boolean {
        if (valueToCheck.isEmpty()) {
            return false
        }
        val currentFormattedChars = valueToCheck.trim { it <= ' ' }.toCharArray()
        val lastIndex = currentFormattedChars.size - 1

        return currentFormattedChars[lastIndex] == ':'
    }

    private fun indentValue(depth: Int): String {
        var indentedValue = ""
        for (j in 0 until depth) {
            indentedValue += " "
        }
        return indentedValue
    }

    companion object {
        private const val DEFAULT_INDENT = 4
    }
}
